odl/solvers/functional/functional.py

```diff
@@ -1268,6 +1333,186 @@ def __str__(self):
+                                 '`subgrad.domain` {}'
+                                 ''.format(self.__functional.domain,
+                                           subgrad.domain))
+            if subgrad.range != self.__functional.domain:
+                raise ValueError('`functional.domain` {} is not the same as '
+                                 '`subgrad.range` {}'
+                                 ''.format(self.__functional.domain,
+                                           subgrad.range))
+            self.__subgrad = subgrad
+
+        self.__initialized = False
+        self.__subgrad_eval = None
+        self.__constant = None
+        self.__bregman_dist = None
+
+        super(BregmanDistance, self).__init__(space=functional.domain,
+                                              linear=False)
+
+    @property
+    def functional(self):
+        """The functional for the Bregman distance."""
+        return self.__functional
+
+    @property
+    def point(self):
+        """The point for the Bregman distance."""
+        return self.__point
+
+    @property
+    def subgrad(self):
+        """The subgradient operator for the Bregman distance."""
+        return self.__subgrad
+
+    @property
+    def initialized(self):
+        """Flag for if the Bregman distance functional is initialized."""
+        return self.__initialized
+
+    def _initialize(self):
+        """Helper function to initialize the functional."""
+
+        self.__subgrad_eval = self.subgrad(self.point)
+        self.__constant = (-self.functional(self.point) +
+                           self.__subgrad_eval.inner(self.point))
+        self.__bregman_dist = FunctionalQuadraticPerturb(
+            self.functional, linear_term=-self.__subgrad_eval,
+            constant=self.__constant)
+
+        self.grad_lipschitz = (self.functional.grad_lipschitz +
+                               self.__subgrad_eval.norm())
+
+        self.__initialized = True
+
+    def _call(self, x):
+        """Return ``self(x)``."""
+        if not self.initialized:
+            self._initialize()
+
+        return self.__bregman_dist(x)
+
+    @property
+    def convex_conj(self):
+        """The convex conjugate"""
+        if not self.initialized:
+            self._initialize()
+
+        return self.__bregman_dist.convex_conj
+
+    @property
+    def proximal(self):
+        """Return the ``proximal factory`` of the functional."""
+        if not self.initialized:
+            self._initialize()
+
+        return self.__bregman_dist.proximal
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional."""
+        if not self.initialized:
+            self._initialize()
+
+        return self.subgrad - ConstantOperator(self.__subgrad_eval)
+
+    def __repr__(self):
+        '''Return ``repr(self)``.'''
+        posargs = [self.functional, self.point]
+        subgrad_op_default = getattr(self.functional, 'gradient', None)
```

Comment:
Functional always implements `gradient` (since it is an `Functional`), so the `getattr` here is redundant

Reply:
I do not think I fully understand how `signature_string` works... 

But even the current implementation can get into trouble. If I uses, e.g., an `IndicatorBox` which I supplied with an appropriate subgradient operator in the creation of the `BregmanDistance` as follows 

```python
>>> import odl
>>> space = odl.uniform_discr(0,1,10)
>>> indic_func = odl.solvers.IndicatorBox(space, 0.2, 0.8)
>>> point = 0.5 * space.one()
>>> subgrad = odl.ZeroOperator(space)
>>> breg = odl.solvers.BregmanDistance(indic_func, point, subgrad)
>>> breg.__repr__()
```

 I get the following output

```
Traceback (most recent call last):

  File "<ipython-input-1-dcb69fc015ba>", line 7, in <module>
    breg.__repr__()

  File "[...]/odl/odl/solvers/functional/functional.py", line 1503, in __repr__
    subgrad_op_default = getattr(self.functional, 'gradient', None)

  File "[...]/odl/odl/solvers/functional/functional.py", line 97, in gradient
    ''.format(self))

NotImplementedError: no gradient implemented for functional IndicatorBox(uniform_discr(0.0, 1.0, 10), 0.2, 0.8)
```

However `self.func_subgrad` will, by the way that `__init__` works, always have a value. Either it is supplied separately or it is set to be the gradient. Thus one could just change to

```python
        optargs = [('func_subgrad', self.func_subgrad, None)]
```

where `None` will never be used, right?

Reply:
This solution should be fine. It will make `repr` always print out the `func_subgrad` even if the user didn't specify it, but the logic is complex enough that you may want to skip it.
If you don't want this behavior, you probably need to do the `gettattr` thing to get the default and use that instead of `None` in `optargs`.